rust/lance/src/dataset/write/commit.rs

```diff
@@ -0,0 +1,339 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use lance_file::version::LanceFileVersion;
+use lance_io::object_store::{ObjectStore, ObjectStoreParams, ObjectStoreRegistry};
+use lance_table::{
+    format::DataStorageFormat,
+    io::commit::{CommitConfig, CommitHandler, ManifestNamingScheme},
+};
+use snafu::{location, Location};
+
+use crate::{
+    dataset::{
+        builder::DatasetBuilder,
+        commit_detached_transaction, commit_new_dataset, commit_transaction,
+        refs::Tags,
+        transaction::{Operation, Transaction},
+        ManifestWriteConfig, ReadParams,
+    },
+    session::Session,
+    Dataset, Error, Result,
+};
+
+use super::{resolve_commit_handler, WriteDestination};
+
+/// Create a new commit from a [`Transaction`].
+///
+/// Transactions can be created using a write method like [`super::InsertBuilder::execute_uncommitted`].
+#[derive(Debug, Clone)]
+pub struct CommitBuilder<'a> {
+    dest: WriteDestination<'a>,
+    use_move_stable_row_ids: Option<bool>,
+    enable_v2_manifest_paths: bool,
+    storage_format: Option<LanceFileVersion>,
+    commit_handler: Option<Arc<dyn CommitHandler>>,
+    store_params: Option<ObjectStoreParams>,
+    object_store_registry: Arc<ObjectStoreRegistry>,
+    object_store: Option<Arc<ObjectStore>>,
+    session: Option<Arc<Session>>,
+    detached: bool,
+    commit_config: CommitConfig,
+}
+
+impl<'a> CommitBuilder<'a> {
+    pub fn new(dest: impl Into<WriteDestination<'a>>) -> Self {
+        Self {
+            dest: dest.into(),
+            use_move_stable_row_ids: None,
+            enable_v2_manifest_paths: false,
+            storage_format: None,
+            commit_handler: None,
+            store_params: None,
+            object_store_registry: Default::default(),
+            object_store: None,
+            session: None,
+            detached: false,
+            commit_config: Default::default(),
+        }
+    }
+
+    /// Whether to use move-stable row ids. This makes the `_rowid` column stable
+    /// after compaction, but not updates.
```

Comment:
even during re-compaction? 
interesting, not updates, will update create new rows with new _rowid?

Reply:
Even during re-compaction, yeah. We don't keep them during updates because that requires invalidating the secondary indices.

You can read more in the design docs:

* [Move-stable row ids](https://www.notion.so/lancedb/Lance-Move-Stable-Row-Ids-7664ef3190de4e3b95f05648101fe8bc?pvs=4)
* [Primary keys](https://www.notion.so/lancedb/Lance-Primary-keys-0a38059cfff44af7b12310cfd45408a1?pvs=4) - when we make row ids stable after updates too

Reply:
Thanks Will! will take a look at the design docs!